### ISSUE TITLE
kubelet memory leaks for housekeeping goroutines keeps leakiness

### ISSUE THREAD

**Issue by attlee-wang** [google/cadvisor#3218]
### What happened?

In my online kubernetes cluster, kubelet memory keeps growing, finally more than 50G, and kill many low-priority processes with memory. 
<img width="1342" alt="image" src="https://user-images.githubusercontent.com/57479557/210139100-e406d2e5-5894-4648-8c60-2936b15db955.png">
I observed goroutines of kubelet are also increasing synchronously with the memory：
<img width="1344" alt="image" src="https://user-images.githubusercontent.com/57479557/210139168-54851a95-378f-4a86-a321-3fdd2e1281b0.png">
I use golang pprof analysis and found that many goroutines stay in [the housekeeping logic](https://github.com/google/cadvisor/blob/92d49a1c4cd1b0c983883fa965406355fe96bcb6/manager/container.go#L446:~:text=if%20!c.housekeepingTick,%7D) 

<div align="center">
  <img src="https://user-images.githubusercontent.com/57479557/210139290-38d32074-700a-4907-a7d0-623e719f8aad.png" width=40% title="pprof housekeeping">
</div>

[housekeeping() goroutines](https://github.com/google/cadvisor/blob/92d49a1c4cd1b0c983883fa965406355fe96bcb6/manager/container.go#L108) have only one exit point, which is [read <-c.stop chan massage](https://github.com/google/cadvisor/blob/92d49a1c4cd1b0c983883fa965406355fe96bcb6/manager/container.go#L446).  But I didn't find anything unusual by checking the kebelet log, so why housekeeping() goroutines do keep growing?

### What did you expect to happen?

Housekeeping() goroutines exit normally and find out why it can't exit normally.

### How can we reproduce it (as minimally and precisely as possible)?

I don't know reproduce it, restart the kubelet and the problem will disappear

### Kubernetes version

<details>

```console
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.5", GitCommit:"---", GitTreeState:"clean", BuildDate:"2020-06-26T03:47:41Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.5", GitCommit:"---", GitTreeState:"clean", BuildDate:"2020-06-26T03:39:24Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}
```

</details>

### Cloud provider

<details>
Internal private cloud platform
</details>

### OS version

<details>

```console
# On Linux:
root@:~# cat /etc/os-release
PRETTY_NAME="Debian GNU/Linux bookworm/sid"
NAME="Debian GNU/Linux"
VERSION_CODENAME=bookworm
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

root@:~# crictl version
Version:  0.1.0
RuntimeName:  containerd
RuntimeVersion:  v1.5.5-9
RuntimeApiVersion:  v1alpha2

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by attlee-wang** [comment]
@pacoxu [pacoxu](https://github.com/pacoxu)  I also noticed issues  [#3014](https://github.com/google/cadvisor/pull/3014), but after my investigation, the reason is different.

I add log to [housekeeping()](https://github.com/google/cadvisor/blob/92d49a1c4cd1b0c983883fa965406355fe96bcb6/manager/container.go#L446:~:text=if%20!c.housekeepingTick,%7D) found that add housekeeping() is much more than stop housekeeping(), but there are only a few containers on my node.

![38D05xyX8S](https://user-images.githubusercontent.com/57479557/211016460-9ccec7b8-111b-444c-9381-3ccfdffccd38.jpg)

Adding logs to the code, I found that the [lastWatched of the failed container is always false](https://github.com/google/cadvisor/blob/92d49a1c4cd1b0c983883fa965406355fe96bcb6/container/raw/watcher.go#L216:~:text=watching%20for%20it.-,lastWatched%2C%20err%20%3A%3D%20self.watcher.RemoveWatch(containerName,%7D,-default%3A)), so it is not sent stop housekeeping(). But I didn't find why [lastWatched is always false]((https://github.com/google/cadvisor/blob/92d49a1c4cd1b0c983883fa965406355fe96bcb6/container/raw/watcher.go#L216:~:text=watching%20for%20it.-,lastWatched%2C%20err%20%3A%3D%20self.watcher.RemoveWatch(containerName,%7D,-default%3A))) of the failed container.  @[pacoxu](https://github.com/pacoxu)  @[iwankgb](https://github.com/iwankgb)

**Comment by attlee-wang** [comment]
> Kubernetes v1.18.5 is using cadvisor v0.35.0.

@pacoxu yes. I sync [# 100326](https://github.com/kubernetes/kubernetes/pull/100326) and [# issues](https://github.com/kubernetes/kubernetes/issues/114751) codes, but problem still exists. It may be that [# 94583](https://github.com/kubernetes/kubernetes/issues/94583) has not been fully fixed.

**Comment by iwankgb** [comment]
@pacoxu is this version of Kubernetes still supported?
